Userland/Libraries/LibGUI/UIDimensions.h

```diff
@@ -0,0 +1,304 @@
+    {
+        for (SpecialDimension v : valid_values) {
+            if (m_value == int(v) || (v == SpecialDimension::Regular && is_regular_value()))
+                return true;
+        }
+        return false;
+    }
+
+    [[nodiscard]] inline bool operator==(UIDimension const& other) const
+    {
+        return m_value == other.m_value;
+    }
+
+    [[nodiscard]] inline UIDimension regular_sum_with_verify(UIDimension const& other) const
+    {
+        VERIFY(is_regular_value() && other.is_regular_value());
+        return UIDimension { m_value + other.m_value };
+    }
+
+    inline void add_to_regular_with_verify(int to_add)
+    {
+        VERIFY(is_regular_value());
+        VERIFY(m_value >= -to_add);
+        m_value += to_add;
+    }
+
+    inline void add_if_regular(int to_add)
+    {
+        if (is_regular_value()) {
+            m_value += to_add;
+        }
+    }
+
+    [[nodiscard]] inline int value_or_zero_if_shrink_with_verify() const
+    {
+        if (m_value >= 0)
+            return m_value;
+        if (m_value == int(SpecialDimension::Shrink))
+            return 0;
+        VERIFY_NOT_REACHED();
+    }
+
+    [[nodiscard]] inline int value_verify_regular() const
+    {
+        VERIFY(is_regular_value());
+        return m_value;
+    }
+
+    [[nodiscard]] AK::JsonValue as_json_value() const
+    {
+        if (is_regular_value())
+            return m_value;
+        else {
+            if (is_shrink())
+                return "shrink";
+            else if (is_grow())
+                return "grow";
+            else if (is_opportunistic_grow())
+                return "opportunistic_grow";
+            else if (is_fit())
+                return "fit";
+            else
+                VERIFY_NOT_REACHED();
+        }
+    }
+
+    [[nodiscard]] static Optional<UIDimension> construct_from_json_value(AK::JsonValue const value)
+    {
+        if (value.is_string()) {
+            String value_literal = value.as_string();
+            if (value_literal == "shrink")
+                return UIDimension { SpecialDimension::Shrink };
+            else if (value_literal == "grow")
+                return UIDimension { SpecialDimension::Grow };
+            else if (value_literal == "opportunistic_grow")
+                return UIDimension { SpecialDimension::OpportunisticGrow };
+            else if (value_literal == "fit")
+                return UIDimension { SpecialDimension::Fit };
+            else
+                return {};
+        } else {
+            int value_int = value.to_i32();
+            if (value_int < 0)
+                return {};
+            return UIDimension(value_int);
+        }
+    }
+};
+
+class UISize : public Gfx::Size<UIDimension> {
+
+public:
+    UISize() = delete;
+
+    UISize(int in_width, int in_height)
+        : Gfx::Size<UIDimension>(in_width, in_height)
+    {
+        VERIFY(width().is_regular_value() && height().is_regular_value());
```

Comment:
This should never happes, no need to VERIFY it at runtime (type system already gaurenties the value must be regular if Gfx::Size keeps the expected contract. (value passed to constructor is returned as expected from the getters)